### ISSUE TITLE
Proposal: API for Call Creation

### ISSUE THREAD

**Issue by naturalwarren** [square/okhttp#3515]
When receiving a 307 we’d like the ability to cancel all inflight requests and retry them with the new hostname received.

Today, we have a small fork of OkHttp that facilitates this by exposing `Call` in `Interceptor.Chain`: 

```java
public interface Interceptor {
  Chain {
    Call call()
  }
}
```

Using an `Interceptor`, we keep track of all inflight calls. When a 307 is received we cancel and retry all of them with the new hostname.

Since the only way to plug into `Call` creation is to subclass `OkHttpClient` or be a `Call.Factory` this suggests the need for a new API similar to Retrofit’s `CallAdapter`. 

Thoughts?

**Comment by swankjesse** [comment]
Is this a request to add a a `call()` method to `Chain` ? I don’t quite follow what ya want here.

**Comment by yschimke** [comment]
FWIW: I've had a similar issue with automatically renewing an OAuth2 token in an application interceptor.  I got around it mostly by having a second OkHttpClient use for auth requests.

1. My application interceptor attaches credentials to request and calls chain.proceed(request)
2. Request fails with 401, interceptor fires off a synchronous Oauth2 refresh request
3. interceptor retries original request in a new call

I'm not sure it is safe to do what I'm doing, create a new call that potentially shares the same dispatcher etc.  So I'm not sure that OkHttp should support it, it has the potential to make it very easy to create a gridlock. But I'm writing a command line client so I can estimate I should be ok.

**Comment by swankjesse** [comment]
@yschimke I think you’ll be safe if you’re doing a synchronous call. Sync calls don’t use the dispatcher and aren’t subject to connection limits.

**Comment by naturalwarren** [comment]
This is addressed by #3557 since we now have an API to track in flight calls.
